src/walker.ts

```diff
@@ -0,0 +1,62 @@
+import { PotentialStep, Pipeline, Conditional, Command, CommandStep } from '.';
+import { getAndCacheDependency } from './conditional';
+
+export type Entity = Pipeline | PotentialStep | Command;
+
+async function walkEntity<T extends Entity>(
+  entity: T,
+  nodeCache: Map<any, any>,
+  mutator?: (entity: Entity) => any,
+): Promise<T> {
+  if (nodeCache.has(entity)) {
+    return nodeCache.get(entity);
+  }
+
+  const cache = new Map();
+
+  if (entity instanceof Conditional) {
+    if (await entity.accept()) {
+      entity = <any>await getAndCacheDependency(cache, entity);
```

Comment:
unwrap the conditional early if it's accepted

Reply:
The other uses call `getAndCacheDependency` before calling `accept`.

Reply:
I think it's ok to unwrap the entity inside the conditional and walk over it, but that should be done unconditionally, and the conditional itself should also be walked over.

That way the mutator can either handle all steps ignoring whether they're conditional or not, or handle conditional steps differently.

Reply:
I think the problem with unconditionally wrapping, is we need to wrap it again after it's walked, I'm not sure if we can since we don't actually know what type of conditional it is.

@joscha do you know if this is possible?

Reply:
Discussed with @joscha (who in turn discussed with @uri-canva ). To simplify for now we will unwrap all conditionals in the walk phase of the graph, if they have been accepted, we will replace the conditional with the guarded step, if they are rejected, we will remove them from the graph.

To make the walk easier, @uri-canva suggested we do this conditional walk before the actual walk, so I've added two extra methods to evaluate all conditionals in the pipeline.